### PR TITLE
Add support for async filters

### PR DESCRIPTION
**Pull request by mauve**
Adds async document, operation, parameter and request body filters.

Fixes: #1661

### CODE REVIEW

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/IDocumentFilter.cs
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/IDocumentFilter.cs:15]:
```suggestion
        Task ApplyAsync(OpenApiDocument swaggerDoc, DocumentFilterContext context);
```

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/IOperationFilter.cs
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/IOperationFilter.cs:15]:
```suggestion
        Task ApplyAsync(OpenApiOperation operation, OperationFilterContext context);
```

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/IParameterFilter.cs
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/IParameterFilter.cs:15]:
```suggestion
        Task ApplyAsync(OpenApiParameter parameter, ParameterFilterContext context);
```

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/IRequestBodyFilter.cs
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/IRequestBodyFilter.cs:15]:
```suggestion
        Task ApplyAsync(OpenApiRequestBody requestBody, RequestBodyFilterContext context);
```

**Comment by mauve** [comment]
Hi @martincostello, Thanks for taking over maintainership of the project ❤️.

Is there a way for me to trigger the checks?

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGenerator.cs
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGenerator.cs:74]: I don't think there's a need to do this - if someone needs a filter that's asynchronous and has others that don't need it, they should be able to implement it with both interfaces, not force them to refactor all their others to make everything match.

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGenerator.cs
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGenerator.cs:128]: This seems to introduce a lot of duplication between the two implementations. Can it be refactored to just make the existing methods async and then iterate over both sets?

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGeneratorOptions.cs
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGeneratorOptions.cs:82]: I don't we need this - they shouldn't be mutually exclusive.

File: test/Swashbuckle.AspNetCore.SwaggerGen.Test/SwaggerGenerator/SwaggerGeneratorAsyncTests.cs
Comment (martincostello) [inline test/Swashbuckle.AspNetCore.SwaggerGen.Test/SwaggerGenerator/SwaggerGeneratorAsyncTests.cs:15]: Just add these tests to the existing test class?

**Comment by martincostello** [comment]
> Is there a way for me to trigger the checks?

No, it's a built-in feature of GitHub Actions for first time contributors to a repo - someone with write access needs to approve them to run until your first contribution is merged.

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGenerator.cs
Comment (mauve) [reply src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGenerator.cs:128]: I was thinking about something like that, but I am not sure how I can call these async methods in the existing non-async code-path?

Don't we need to support `ISwaggerProvider.GetSwagger()` still?

File: test/Swashbuckle.AspNetCore.SwaggerGen.Test/SwaggerGenerator/SwaggerGeneratorAsyncTests.cs
Comment (mauve) [reply test/Swashbuckle.AspNetCore.SwaggerGen.Test/SwaggerGenerator/SwaggerGeneratorAsyncTests.cs:15]: Will do

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGenerator.cs
Comment (martincostello) [reply src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGenerator.cs:128]: I only looked at the diff, I didn't go digging through all the code. Can you point me at the bit where you're looking?

We would need to keep the existing methods/support though, yes.

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGeneratorOptions.cs
Comment (martincostello) [reply src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGeneratorOptions.cs:82]: Ideally people should be able to use both if we can design it that way.

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGenerator.cs
Comment (mauve) [reply src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGenerator.cs:74]: I don't think that is what this code will do, at least what I meant that it does is that it fails if you use an async filter with the sync `GetSwagger()`.

I am not 100% sure when `GetSwagger()` is called instead of `GetSwaggerAsync()` the middleware seems to prefer the async version of the sync version.

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGeneratorOptions.cs
Comment (mauve) [reply src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGeneratorOptions.cs:82]: I don't feel super-strongly about this, just let me know which option you prefer:

1. GetSwagger() silently ignores async filters
2. GetSwagger() throws error if async filters are used

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGenerator.cs
Comment (mauve) [reply src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGenerator.cs]: I figured out a way to remove the duplication.

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGeneratorOptions.cs
Comment (mauve) [reply src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGeneratorOptions.cs:82]: > Ideally people should be able to use both if we can design it that way.

Sorry, I missed this comment before my last answer. What are you referring to with `both`? `GetSwagger()` and `GetSwaggerAsync()` or async and non-async filters?

Comment (martincostello) [reply src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGeneratorOptions.cs:82]: Just the filters - if there's an async middleware, then it could use both implementations, where as a sync middleware/interface/whatever would only be able to use the sync ones - in that case it should probably just ignore the async ones entirely.

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGenerator.cs
Comment (martincostello) [reply src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGenerator.cs:75]: When I have time to clone the code locally and dig through it in an IDE I can take a better look at what's the best way to go about this to try and make it "just work" for the async case.

If we were to make breaking changes in the future we'd probably remove the sync versions if we also have async.

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGeneratorOptions.cs
Comment (mauve) [reply src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGeneratorOptions.cs:82]: Ok, I will implement that.

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGenerator.cs
Comment (mauve) [reply src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGenerator.cs]: I have now removed this check (and the corresponding property on SwaggerGeneratorOptions)

**Comment by mauve** [comment]
@martincostello The tests are now green :) Let me know if you want any further changes.

**Comment by martincostello** [comment]
Thanks - I'll take a look at this in more detail tomorrow 👍

**Comment by mauve** [comment]
@martincostello Hi, have you had chance to look at it?

**Comment by martincostello** [comment]
Sorry, not yet, but it's on my TODO list.

**Comment by martincostello** [comment]
Hi @mauve - I checked out your pull request locally and had a look around the code. I've done some refactoring in this commit that aims to reduce the duplication in `SwaggerGenerator`: https://github.com/martincostello/Swashbuckle.AspNetCore/commit/01cf4d585f5416649da9dc2fe3b6eb97ca6e3594

What are your thoughts on the approach?

**Comment by mauve** [comment]
Great! Unsure what the process should be but I just cherry-picked your commit to the top of my PR.

**Comment by martincostello** [comment]
Thanks - I'll take a proper look at this again either later this afternoon, or more likely, tomorrow 👍

File: src/Swashbuckle.AspNetCore.Swagger/IAsyncSwaggerProvider.cs
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.Swagger/IAsyncSwaggerProvider.cs:1]: Sorry, my bad here - I thought this interface was new as part of your PR, not something we already had 😅

This would be a breaking change, so the `cancellationToken` parameter needs to be removed.

That _might_ then make the other `CancellationToken` additions to the new interfaces pointless, unless the `HttpContext` happens to be already present in what's available within `SwaggerGenerator`, in which case it can pass through `HttpContext.RequestAborted`.

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGeneratorOptions.cs
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGeneratorOptions.cs:72]:
```suggestion
        public IList<IRequestBodyAsyncFilter> RequestBodyAsyncFilters { get; set; }
```

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGeneratorOptions.cs
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGeneratorOptions.cs:74]:
```suggestion
        public IList<IOperationFilter> OperationFilters { get; set; }
```

Comment (mauve) [reply src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGeneratorOptions.cs]: I have changed this aswell, however this is an existing property, isn't that technically an API break?

**Comment by mauve** [comment]
@martincostello I have removed the CancellationToken from GetSwaggerAsync, however I left them in the newly added interfaces so that we don't need to break that interface when we introduce CancellationToken again.

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGeneratorOptions.cs
Comment (martincostello) [reply src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGeneratorOptions.cs]: I must have commented on the wrong line - the new properties should use `IList<T>` but anything already existing should be left alone.

File: src/Swashbuckle.AspNetCore.Swagger/IAsyncSwaggerProvider.cs
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.Swagger/IAsyncSwaggerProvider.cs:1]: This can be removed.

```suggestion
```

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGeneratorOptions.cs
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGeneratorOptions.cs:74]:
```suggestion
        public List<IOperationFilter> OperationFilters { get; set; }
```

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGeneratorOptions.cs
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGeneratorOptions.cs:68]:
```suggestion
        public List<IParameterAsyncFilter> ParameterAsyncFilters { get; set; }
```

**Comment by mauve** [comment]
I accepted all your suggestions @martincostello

File: src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGeneratorOptions.cs
Comment (martincostello) [inline src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGeneratorOptions.cs:70]: These two are the wrong way around:

```suggestion
        public IList<IParameterAsyncFilter> ParameterAsyncFilters { get; set; }

        public List<IRequestBodyFilter> RequestBodyFilters { get; set; }
```

**Review by martincostello — APPROVED** [review]

**Comment by mauve** [comment]
@martincostello Thanks 🎉 I am very happy have made my first contribution :)

**Comment by mauve** [comment]
Hi @martincostello do you have any plans to create a new release? Is there anything I can assist with in that case?

**Comment by martincostello** [comment]
There's been a drip-feed of changes happening, so I've been waiting for things to quieten down.

**Pull request merged**
